modules/cudf/src/data_frame.ts

```diff
@@ -70,6 +70,93 @@ function _invokeIfNumericSeries<P extends keyof T, T extends TypeMap, R extends
   return Series.new(series._col as Column<R>);
 }
 
+function _concat<DFs extends DataFrame[], T extends TypeMap>(...dfs: DFs): DataFrame {
+  const all_column_names = [...dfs
+                              .reduce((namesMap, df) => df.names.reduce(
+                                        (namesMap, name) => namesMap.set(name, true), namesMap),
+                                      new Map<string, boolean>())
+                              .keys()];
+
+  /**
+   * Array<Array<(Column|null)>> -- If a DF has a Column for a given name, it will be in the list
+   * otherwise there will be a null in that slot. For example:
+   * ```
+   * concat(new DataFrame({a, b}), new DataFrame({b, c}))
+   *
+   * columnsPerDF == [
+   *  [dfs[0].get("a"), dfs[0].get("b"),            null],
+   *  [           null, dfs[1].get("b"), dfs[1].get("c")]
+   * ]
+   * ```
+   */
+  const columns_per_df: any[][] =
+    dfs.map((df) => all_column_names.map((name) => df.has(name) ? df.get(name)._col : null));
+
+  const num_of_cols = columns_per_df[0].length;
+  const num_of_rows = columns_per_df.length;
+
+  /**
+   * Array<[number, DataType]> -- Find the first non null dtype in each column, save the column
+   * index and dtype in a tuple.
+   * ```
+   * [[0, Float64], [1, Int32], [2, Float64]],
+   * ```
+   */
+  const first_non_null_dtype: [number, DataType][] = Array(num_of_cols).fill(null);
+  first_non_null_dtype.forEach((_, col_idx) => {
+    for (let row_idx = 0; row_idx < num_of_rows; ++row_idx) {
+      const col = columns_per_df[row_idx][col_idx];
+      if (col !== null) { first_non_null_dtype[col_idx] = [row_idx, col.type]; }
+    }
+  });
+
+  /**
+   * Find the common dtype in each column.
+   * ```
+   * [Float64, Int32, Float64]
+   * ```
+   */
+  const common_dtypes: DataType[] = first_non_null_dtype.map((tuple) => { return tuple[1]; });
+  first_non_null_dtype.forEach((tuple, col_idx) => {
+    const first_non_null_dtype_idx = tuple[0];
+    const start_idx                = first_non_null_dtype_idx + 1;
+    for (let row_idx = start_idx; row_idx < num_of_rows; ++row_idx) {
+      const first_non_null_dtype = tuple[1];
+      const col                  = columns_per_df[row_idx][col_idx];
+      if (col !== null) { common_dtypes[col_idx] = findCommonType(first_non_null_dtype, col.type); }
+    }
+  });
+
+  /**
+   * If any columns are null, create an empty column with type of common dtype
+   * Otherwise cast the column using the common dtype (if it isn't already the correct type).
+   */
+  common_dtypes.forEach((common_dtype, col_idx) => {
+    for (let row_idx = 0; row_idx < num_of_rows; ++row_idx) {
+      const col = columns_per_df[row_idx][col_idx];
+      if (col === null) {
+        const empty_column_length =
+          columns_per_df.find((row) => row !== null)?.find((col) => col !== null).length;
```

Comment:
Could be the case that `find` returns a []. Creating an array of size `0`, and then throwing once the tables are being constructed. Not sure if we want to handle specifically with a throw.